In [3]:
from graph import Graph
from tqdm import tqdm
import numpy as np
import time
import random
from collections import deque
import matplotlib.pyplot as plt

In [4]:
page_graph = Graph()
page_graph_bidirectional = Graph()

with open('web-Google.txt', 'r') as file:
    for l in file:
        if "# FromNodeId	ToNodeId" in l:
            break
    for l in tqdm(file,total =5105039):
        if not l:
            break
        edge = tuple(int(v.replace("\n", "").replace("\t", "")) for v in l.split("\t"))
        for v in edge:
            if not page_graph.vertex_exists(v):
                page_graph.add_vertex(str(v))
                page_graph_bidirectional.add_vertex(str(v))
        page_graph.add_edge(str(edge[0]), str(edge[1]))
        page_graph_bidirectional.add_edge(str(edge[0]), str(edge[1]))
        page_graph_bidirectional.add_edge(str(edge[1]), str(edge[0]))


  0%|          | 0/5105039 [00:00<?, ?it/s]

100%|██████████| 5105039/5105039 [00:46<00:00, 110050.31it/s]


In [5]:
def bfs(graph, start_vertex):
    visited = set()
    queue = deque([start_vertex])
    while queue:
        vertex = queue.popleft()
        if vertex not in visited:
            visited.add(vertex)
            queue.extend(set(graph.get_neighbors(vertex)) - visited)
    return visited

def dfs(graph, start_vertex):
    visited = set()
    stack = deque([start_vertex])
    while stack:
        vertex = stack.pop()
        if vertex not in visited:
            visited.add(vertex)
            stack.extend(set(graph.get_neighbors(vertex)) - visited)
    return visited

def connected_components(graph):
    start_time = time.time()
    visited = set()
    components = []
    for vertex in tqdm(graph.get_vertexes(),total=len(graph.get_vertexes())):
        if vertex not in visited:
            component = dfs(graph, vertex)
            visited.update(component)
            components.append(component)
    print(f"Execution time: {time.time() - start_time}")
    return components

def estimacion_caminos_minimos(graph, cantidad_de_muestras):
    start_time = time.time()
    for i in tqdm(range(cantidad_de_muestras),total=cantidad_de_muestras):
        v1 = random.choice(list(graph._graph.keys()))
        bfs(graph, v1)
    total_time = time.time() - start_time
    print(f"Execution time: {total_time}")
    hours, rem = divmod((total_time / cantidad_de_muestras * 875713), 3600)
    minutes, seconds = divmod(rem, 60)
    days, hours = divmod(hours, 24)
    return f"{int(days)} dias y {int(hours)}:{int(minutes)}:{int(seconds)}"
    
def contar_triangulos(graph):
    start_time = time.time()
    count = 0
    for vertex in tqdm(graph.get_vertexes(),total=len(graph.get_vertexes())):
        for neighbor in graph.get_neighbors(vertex):
            count += len(set(graph.get_neighbors(neighbor)) & set(graph.get_vertexes_from(vertex)))
    print(f"Execution time: {time.time() - start_time}")
    return count

def diametro_estimacion(graph, cantidad_de_muestras):
    start_time = time.time()
    max_distance = 0
    for _ in tqdm(range(cantidad_de_muestras),total=cantidad_de_muestras):
        v1 = random.choice(list(graph.get_vertexes()))
        distances = {v1: 0}
        queue = deque([v1])
        while queue:
            vertex = queue.popleft()
            for neighbor in graph.get_neighbors(vertex):
                if neighbor not in distances:
                    distances[neighbor] = distances[vertex] + 1
                    max_distance = max(max_distance, distances[neighbor])
                    queue.append(neighbor)
    total_time = time.time() - start_time
    print(f"Execution time: {total_time}")
    return max_distance

def diametro_estimacion_dfs(graph, cantidad_de_muestras):
    start_time = time.time()
    max_distance = 0
    for _ in tqdm(range(cantidad_de_muestras),total=cantidad_de_muestras):
        v1 = random.choice(list(graph.get_vertexes()))
        distances = {v1: 0}
        stack = deque([v1])
        while stack:
            vertex = stack.pop()
            for neighbor in graph.get_neighbors(vertex):
                if neighbor not in distances:
                    distances[neighbor] = distances[vertex] + 1
                    max_distance = max(max_distance, distances[neighbor])
                    stack.append(neighbor)
    total_time = time.time() - start_time
    print(f"Execution time: {total_time}")
    return max_distance
    
def pagerank(graph, iterations=100, damping_factor=0.85, epsilon=1e-6):
    start_time = time.time()
    n = len(graph.get_vertexes())
    ranks = {v: 1.0 / n for v in graph.get_vertexes()}
    for _ in tqdm(range(iterations), total=iterations):
        new_ranks = {}
        for vertex in graph.get_vertexes():
            new_ranks[vertex] = (1 - damping_factor) / n
            for neighbor in graph.get_vertexes_from(vertex):
                new_ranks[vertex] += damping_factor * ranks[neighbor] / len(graph.get_neighbors(neighbor))
        ranks = new_ranks
        if all(abs(ranks[v] - 1.0 / n) < epsilon for v in graph.get_vertexes()):
            break
    print(f"Execution time: {time.time() - start_time}")
    return ranks or {} 

def circunferencia_estimacion(graph, cantidad_de_muestras):
    start_time = time.time()
    max_distance = 0
    for _ in tqdm(range(cantidad_de_muestras),total=cantidad_de_muestras):
        v1 = random.choice(list(graph.get_vertexes()))
        distances = {v1: 0}
        queue = deque([v1])
        while queue:
            vertex = queue.popleft()
            for neighbor in graph.get_neighbors(vertex):
                if neighbor not in distances:
                    distances[neighbor] = distances[vertex] + 1
                    max_distance = max(max_distance, distances[neighbor])
                    queue.append(neighbor)
    total_time = time.time() - start_time
    print(f"Execution time: {total_time}")
    return max_distance

1) ¿Cuál es el tamaño de la componente conexa más grande? ¿Cuántas componentes conexas hay?

In [5]:
components = connected_components(page_graph_bidirectional)
print(f"Number of connected components: {len(components)}")
largest_component = max(components, key=len)
print(f"Size of the largest connected component: {len(largest_component)}")

100%|██████████| 875713/875713 [00:03<00:00, 220631.60it/s]

Execution time: 4.3183276653289795
Number of connected components: 2746
Size of the largest connected component: 855802


Cantidad de componentes conexos: 2746

Tamaño del componente conexo mas grande: 855802

2) Calcular el camino mínimo de todos con todos. ¿En cuanto tiempo lo puede hacer?¿Qué orden tiene el algoritmo? En caso de no alcanzarle el tiempo, estime cuanto tiempo le llevaría."


In [ ]:
print("Podemos obtener todos los caminos minimos en aprox:" ,estimacion_caminos_minimos(page_graph, 100))

2) Calcular el camino mínimo de todos con todos. ¿En cuanto tiempo lo puede hacer?¿Qué orden tiene el algoritmo? En caso de no alcanzarle el tiempo, estime cuanto tiempo le llevaría.


100%|██████████| 100/100 [02:03<00:00,  1.23s/it]

Execution time: 123.32063460350037
Podemos obtener todos los caminos minimos en aprox: 12 dias y 11:58:54


Podemos obtener todos los caminos minimos en aprox: 12 dias y 11:58:54

3) En un grafo un triángulo es una conexión entre 3 vértices A, B y C:
    A está conectado con B
    B está conectado con C
    C está conectado con A
¿Cuántos triángulos tiene el grafo?

In [39]:
print("La cantidad de triangulos es:",contar_triangulos(page_graph))

100%|██████████| 875713/875713 [00:22<00:00, 38358.93it/s]

Execution time: 23.197608709335327
La cantidad de triangulos es: 11669313


La cantidad de triangulos es: 11669313

4) Utilice el punto 2 para calcular el diámetro del grafo.

In [1]:
print("El diametro es:",diametro_estimacion(page_graph, 100))
print("El diametro es:",diametro_estimacion_dfs(page_graph, 100))

NameError: name 'diametro_estimacion' is not defined

El diametro es: 40

5) Google inventó un algoritmo llamado PageRank que le permitía saber qué páginas eran más confiables según que tanto eran referenciadas. PageRank consiste en hacer muchos random walks a lo largo del grafo y contar cuántas veces aparece cada vértice. Los vértices que más aparecen son los de mayor PageRank. Calcule el PageRank de los vértices del grafo.

In [ ]:
ranks = pagerank(page_graph)
sorted_ranks = sorted(ranks.items(), key=lambda x: x[1], reverse=True)[:10]
for node, rank in sorted_ranks:
    print(f"Node: {node}, PageRank: {rank}")

100%|██████████| 100/100 [17:13<00:00, 10.34s/it]


Execution time: 1034.273675441742
Node: 597621, PageRank: 0.0006443556238888784
Node: 41909, PageRank: 0.0006425463534843847
Node: 163075, PageRank: 0.000630599334891639
Node: 537039, PageRank: 0.0006269911082288292
Node: 384666, PageRank: 0.0005489064380190975
Node: 504140, PageRank: 0.0005337160511396311
Node: 486980, PageRank: 0.0005056909681078064
Node: 605856, PageRank: 0.0005008184678214055
Node: 32163, PageRank: 0.0004970632002167673
Node: 558791, PageRank: 0.0004947013146397534


top nodos: 597621, 41909 y 163075

6) La circunferencia del grafo es el largo del ciclo más largo. ¿Cuál es la circunferencia del grafo?

In [48]:
print("La circunferencia es:",circunferencia_estimacion(page_graph, 200))

100%|██████████| 200/200 [03:45<00:00,  1.13s/it]

Execution time: 225.0674114227295
La circunferencia es: 40


La circunferencia es: 38

In [1]:
import time
from collections import deque

def has_cycle_of_size(graph, size, max_time=1000):
    start_time = time.time()
    for start_vertex in graph.get_vertexes():
        stack = deque([(start_vertex, 0, start_vertex)])  # Añadir el tercer elemento para rastrear el vértice inicial
        visited = {start_vertex: 0}  # Llevar un diccionario con la profundidad

        while stack:
            vertex, depth, parent = stack.pop()

            if depth == size:
                if start_vertex in graph.get_neighbors(vertex):
                    return True
                continue

            for neighbor in graph.get_neighbors(vertex):
                if neighbor not in visited:
                    visited[neighbor] = depth + 1
                    stack.append((neighbor, depth + 1, start_vertex))
                elif visited[neighbor] == depth - 1 and neighbor != parent:
                    return True

            if time.time() - start_time > max_time:
                return False

    return False

def max_clique(graph):
    start_time = time.time()
    maximo = 1000
    minimo = 0
    max_clique = 0
    while maximo - minimo > 1:
        if has_cycle_of_size(graph, (maximo + minimo) // 2):
            max_clique = (maximo + minimo) // 2
            minimo = max_clique
            print(f"Max clique: {max_clique}")
        else:
            maximo = (maximo + minimo) // 2
    print(f"Execution time: {time.time() - start_time}")
    return max_clique


In [6]:
print("El tamaño de la clique maxima es:",max_clique(page_graph))

Max clique: 500
Max clique: 750
Max clique: 875
Max clique: 937
Max clique: 968
Max clique: 984
Max clique: 992
Max clique: 996
Max clique: 998
Max clique: 999
Execution time: 2.056002378463745
El tamaño de la clique maxima es: 999
